File: researchflow-production-main/services/worker/src/bridges/ai_router_bridge.py
"""
AI Router Bridge - Python Client

Python wrapper for communicating with the TypeScript AI Router Bridge.
This allows LangGraph agents to make LLM calls through the orchestrator.

Features:
- Async HTTP client for bridge communication
- Automatic model tier selection
- PHI compliance enforcement
- Cost tracking and budget management
- Error handling with retries
- Streaming support (future)
"""

import asyncio
import json
import logging
from typing import Dict, Any, List, Optional, AsyncGenerator
from dataclasses import dataclass, asdict
from enum import Enum
import httpx
from httpx import AsyncClient, TimeoutException, HTTPStatusError

logger = logging.getLogger(__name__)


class ModelTier(Enum):
    """Model tier options"""
    ECONOMY = "ECONOMY"
    STANDARD = "STANDARD"
    PREMIUM = "PREMIUM"


class GovernanceMode(Enum):
    """Governance mode options"""
    DEMO = "DEMO"
    LIVE = "LIVE"
    STANDBY = "STANDBY"


@dataclass
class ModelOptions:
    """Options for LLM model selection"""
    task_type: str
    stage_id: Optional[int] = None
    model_tier: Optional[ModelTier] = None
    governance_mode: Optional[GovernanceMode] = None
    require_phi_compliance: Optional[bool] = None
    budget_limit: Optional[float] = None
    max_tokens: Optional[int] = None
    temperature: Optional[float] = None
    stream_response: Optional[bool] = None
    
    def to_dict(self) -> Dict[str, Any]:
        """Convert to dictionary, handling enums"""
        result = {}
        for key, value in asdict(self).items():
            if value is not None:
                if isinstance(value, Enum):
                    result[key] = value.value
                else:
                    result[key] = value
        return result


@dataclass
class TokenUsage:
    """Token usage statistics"""
    total_tokens: int
    prompt_tokens: int
    completion_tokens: int


@dataclass
class CostBreakdown:
    """Cost breakdown"""
    total: float
    input: float
    output: float


@dataclass
class LLMResponse:
    """Response from LLM call"""
    content: str
    usage: TokenUsage
    cost: CostBreakdown
    model: str
    tier: str
    finish_reason: str
    metadata: Optional[Dict[str, Any]] = None
    
    @classmethod
    def from_dict(cls, data: Dict[str, Any]) -> 'LLMResponse':
        """Create from dictionary response"""
        return cls(
            content=data['content'],
            usage=TokenUsage(**data['usage']),
            cost=CostBreakdown(**data['cost']),
            model=data['model'],
            tier=data['tier'],
            finish_reason=data['finishReason'],
            metadata=data.get('metadata')
        )


@dataclass
class BridgeConfig:
    """Configuration for AI Router Bridge"""
    orchestrator_url: str = "http://localhost:3001"
    default_tier: ModelTier = ModelTier.STANDARD
    phi_compliant_only: bool = True
    max_retries: int = 3
    timeout_seconds: int = 30
    cost_tracking_enabled: bool = True
    streaming_enabled: bool = False


class AIRouterBridgeError(Exception):
    """Base exception for bridge errors"""
    def __init__(self, message: str, code: str = None, details: Any = None):
        super().__init__(message)
        self.code = code
        self.details = details


class RoutingError(AIRouterBridgeError):
    """Error in model routing"""
    pass


class LLMCallError(AIRouterBridgeError):
    """Error in LLM call"""
    pass


class BudgetExceededError(AIRouterBridgeError):
    """Budget limit exceeded"""
    pass


class PHIComplianceError(AIRouterBridgeError):
    """PHI compliance violation"""
    pass


class AIRouterBridge:
    """
    Python client for AI Router Bridge
    
    Handles communication between LangGraph agents and the
    TypeScript AI Router service.
    """
    
    def __init__(self, config: Optional[BridgeConfig] = None):
        """Initialize bridge with configuration"""
        self.config = config or BridgeConfig()
        self.client: Optional[AsyncClient] = None
        self.total_cost = 0.0
        self.request_count = 0
        
    async def __aenter__(self):
        """Async context manager entry"""
        await self.connect()
        return self
        
    async def __aexit__(self, exc_type, exc_val, exc_tb):
        """Async context manager exit"""
        await self.close()
    
    async def connect(self):
        """Initialize HTTP client"""
        if self.client is None:
            timeout = httpx.Timeout(self.config.timeout_seconds)
            self.client = AsyncClient(timeout=timeout)
            
    async def close(self):
        """Close HTTP client"""
        if self.client is not None:
            await self.client.aclose()
            self.client = None
    
    async def invoke(
        self, 
        prompt: str, 
        options: Optional[ModelOptions] = None
    ) -> LLMResponse:
        """
        Single LLM invocation
        
        Args:
            prompt: The prompt to send to the LLM
            options: Model options (defaults to general task)
            
        Returns:
            LLM response with content, usage, and cost
        """
        if options is None:
            options = ModelOptions(task_type="general")
            
        await self.connect()
        
        # Convert options to request format
        request_data = {
            "prompt": prompt,
            "options": options.to_dict()
        }
        
        last_error = None
        
        for attempt in range(1, self.config.max_retries + 1):
            try:
                response = await self.client.post(
                    f"{self.config.orchestrator_url}/api/ai-bridge/invoke",
                    json=request_data
                )
                response.raise_for_status()
                
                # Parse response
                response_data = response.json()
                llm_response = LLMResponse.from_dict(response_data)
                
                # Track costs
                if self.config.cost_tracking_enabled:
                    self.total_cost += llm_response.cost.total
                    self.request_count += 1
                
                logger.info(
                    f"LLM call successful",
                    extra={
                        "task_type": options.task_type,
                        "model": llm_response.model,
                        "tier": llm_response.tier,
                        "tokens": llm_response.usage.total_tokens,
                        "cost": llm_response.cost.total,
                        "attempt": attempt
                    }
                )
                
                return llm_response
                
            except HTTPStatusError as e:
                last_error = e
                
                if e.response.status_code == 400:
                    error_data = e.response.json() if e.response.headers.get('content-type', '').startswith('application/json') else {}
                    error_code = error_data.get('error', 'VALIDATION_ERROR')
                    
                    if error_code == 'BUDGET_INSUFFICIENT':
                        raise BudgetExceededError(
                            error_data.get('message', 'Budget exceeded'),
                            details=error_data
                        )
                    elif error_code == 'PHI_COMPLIANCE_VIOLATION':
                        raise PHIComplianceError(
                            error_data.get('message', 'PHI compliance violation'),
                            details=error_data
                        )
                    else:
                        raise LLMCallError(
                            f"LLM call failed: {error_data.get('message', str(e))}",
                            code=error_code,
                            details=error_data
                        )
                        
                # Don't retry client errors (4xx)
                if 400 <= e.response.status_code < 500:
                    raise LLMCallError(f"Client error: {e}", details=e.response.text)
                
                # Server errors (5xx) - retry with backoff
                if attempt < self.config.max_retries:
                    wait_time = min(2 ** attempt, 60)  # Exponential backoff, max 60s
                    logger.warning(
                        f"LLM call attempt {attempt} failed, retrying in {wait_time}s: {e}"
                    )
                    await asyncio.sleep(wait_time)
                    
            except (TimeoutException, httpx.ConnectError, httpx.RemoteProtocolError) as e:
                last_error = e
                
                if attempt < self.config.max_retries:
                    wait_time = min(2 ** attempt, 60)
                    logger.warning(
                        f"LLM call attempt {attempt} failed, retrying in {wait_time}s: {e}"
                    )
                    await asyncio.sleep(wait_time)
                    
            except Exception as e:
                # Unexpected errors - don't retry
                raise LLMCallError(f"Unexpected error: {e}", details=str(e))
        
        # All retries exhausted
        raise LLMCallError(
            f"Max retries ({self.config.max_retries}) exceeded",
            details=str(last_error)
        )
    
    async def batch_invoke(
        self, 
        prompts: List[str], 
        options: Optional[ModelOptions] = None
    ) -> List[LLMResponse]:
        """
        Batch LLM invocation
        
        Args:
            prompts: List of prompts to process
            options: Model options for all prompts
            
        Returns:
            List of LLM responses
        """
        if options is None:
            options = ModelOptions(task_type="general")
            
        await self.connect()
        
        request_data = {
            "prompts": prompts,
            "options": options.to_dict()
        }
        
        response = await self.client.post(
            f"{self.config.orchestrator_url}/api/ai-bridge/batch",
            json=request_data
        )
        response.raise_for_status()
        
        # Parse batch response
        response_data = response.json()
        results = []
        
        for item in response_data.get('responses', []):
            try:
                results.append(LLMResponse.from_dict(item))
            except Exception as e:
                logger.warning(f"Failed to parse batch response item: {e}")
                # Add error response
                results.append(LLMResponse(
                    content="Error: Failed to parse response",
                    usage=TokenUsage(0, 0, 0),
                    cost=CostBreakdown(0, 0, 0),
                    model="unknown",
                    tier="unknown",
                    finish_reason="error",
                    metadata={"parse_error": str(e)}
                ))
        
        # Update cost tracking
        if self.config.cost_tracking_enabled:
            batch_cost = sum(r.cost.total for r in results)
            self.total_cost += batch_cost
            self.request_count += len(prompts)
        
        return results
    
    async def stream_invoke(
        self, 
        prompt: str, 
        options: Optional[ModelOptions] = None
    ) -> AsyncGenerator[str, None]:
        """
        Streaming LLM invocation
        
        Args:
            prompt: The prompt to send
            options: Model options
            
        Yields:
            Streaming content chunks
        """
        if not self.config.streaming_enabled:
            raise AIRouterBridgeError("Streaming is not enabled")
            
        if options is None:
            options = ModelOptions(task_type="general", stream_response=True)
        else:
            options.stream_response = True
            
        await self.connect()
        
        request_data = {
            "prompt": prompt,
            "options": options.to_dict()
        }
        
        async with self.client.stream(
            "POST",
            f"{self.config.orchestrator_url}/api/ai-bridge/stream",
            json=request_data
        ) as response:
            response.raise_for_status()
            
            content_chunks = []
            
            async for chunk in response.aiter_text():
                if chunk.startswith("data: "):
                    data_str = chunk[6:]  # Remove "data: " prefix
                    
                    if data_str.strip() == "[DONE]":
                        break
                        
                    try:
                        data = json.loads(data_str)
                        if "content" in data:
                            yield data["content"]
                    except json.JSONDecodeError:
                        # Skip invalid JSON chunks
                        continue
    
    async def health_check(self) -> Dict[str, Any]:
        """
        Check bridge and orchestrator health
        
        Returns:
            Health status dictionary
        """
        await self.connect()
        
        try:
            response = await self.client.get(
                f"{self.config.orchestrator_url}/health",
                timeout=5.0
            )
            
            if response.status_code == 200:
                health_data = response.json()
                return {
                    "status": "healthy",
                    "orchestrator_reachable": True,
                    "ai_router_available": health_data.get("services", {}).get("ai_router") == "healthy",
                    "latency_ms": response.elapsed.total_seconds() * 1000
                }
            else:
                return {
                    "status": "unhealthy",
                    "orchestrator_reachable": True,
                    "ai_router_available": False,
                    "latency_ms": response.elapsed.total_seconds() * 1000,
                    "error": f"HTTP {response.status_code}"
                }
                
        except Exception as e:
            return {
                "status": "unhealthy",
                "orchestrator_reachable": False,
                "ai_router_available": False,
                "latency_ms": None,
                "error": str(e)
            }
    
    def get_cost_stats(self) -> Dict[str, float]:
        """
        Get cost tracking statistics
        
        Returns:
            Dictionary with cost statistics
        """
        return {
            "total_cost": self.total_cost,
            "request_count": self.request_count,
            "average_cost_per_request": (
                self.total_cost / self.request_count 
                if self.request_count > 0 
                else 0.0
            )
        }
    
    def reset_cost_tracking(self) -> None:
        """Reset cost tracking counters"""
        self.total_cost = 0.0
        self.request_count = 0


# Factory function
def create_ai_router_bridge(
    orchestrator_url: str = None,
    **kwargs
) -> AIRouterBridge:
    """
    Factory function to create AI Router Bridge
    
    Args:
        orchestrator_url: URL of the orchestrator service
        **kwargs: Additional config options
        
    Returns:
        Configured AIRouterBridge instance
    """
    import os
    
    config = BridgeConfig(
        orchestrator_url=orchestrator_url or os.getenv("AI_ROUTER_URL", "http://localhost:3001"),
        default_tier=ModelTier(os.getenv("AI_DEFAULT_TIER", "STANDARD")),
        phi_compliant_only=os.getenv("PHI_COMPLIANT_ONLY", "true").lower() == "true",
        max_retries=int(os.getenv("AI_MAX_RETRIES", "3")),
        timeout_seconds=int(os.getenv("AI_TIMEOUT_SECONDS", "30")),
        cost_tracking_enabled=os.getenv("AI_COST_TRACKING", "true").lower() == "true",
        streaming_enabled=os.getenv("AI_STREAMING_ENABLED", "false").lower() == "true",
        **kwargs
    )
    
    return AIRouterBridge(config)


# Context manager for easy usage
class AIRouterBridgeContext:
    """Context manager for AI Router Bridge"""
    
    def __init__(self, **kwargs):
        self.bridge = create_ai_router_bridge(**kwargs)
    
    async def __aenter__(self) -> AIRouterBridge:
        await self.bridge.connect()
        return self.bridge
    
    async def __aexit__(self, exc_type, exc_val, exc_tb):
        await self.bridge.close()


# Convenience function
async def invoke_llm(
    prompt: str,
    task_type: str = "general",
    model_tier: ModelTier = None,
    governance_mode: GovernanceMode = None,
    **kwargs
) -> LLMResponse:
    """
    Convenience function for single LLM calls
    
    Args:
        prompt: The prompt to send
        task_type: Type of task
        model_tier: Model tier to use
        governance_mode: Governance mode
        **kwargs: Additional options
        
    Returns:
        LLM response
    """
    options = ModelOptions(
        task_type=task_type,
        model_tier=model_tier,
        governance_mode=governance_mode,
        **kwargs
    )
    
    async with AIRouterBridgeContext() as bridge:
        return await bridge.invoke(prompt, options)


if __name__ == "__main__":
    # Test the bridge
    async def test_bridge():
        async with AIRouterBridgeContext() as bridge:
            # Health check
            health = await bridge.health_check()
            print(f"Health check: {health}")
            
            # Simple test call
            if health["status"] == "healthy":
                try:
                    response = await bridge.invoke(
                        "What is the capital of France?",
                        ModelOptions(task_type="general")
                    )
                    print(f"Response: {response.content[:100]}...")
                    print(f"Cost: ${response.cost.total:.4f}")
                except Exception as e:
                    print(f"Test call failed: {e}")
            else:
                print("Orchestrator not healthy, skipping test call")
    
    asyncio.run(test_bridge())

File: researchflow-production-main/services/orchestrator/src/bridges/ai-router-bridge.ts
/**
 * AI Router Bridge for LangGraph Agent Integration
 *
 * This bridge connects LangGraph agents (Python worker) to the
 * AI Router (TypeScript orchestrator) for LLM calls.
 *
 * Features:
 * - Model tier selection (Economy/Standard/Premium)
 * - PHI-compliant model routing
 * - Cost tracking and budget management
 * - Streaming and batch support
 * - Error handling with fallbacks
 */

import axios from 'axios';
import * as z from 'zod';

// Validation schemas — single source of truth for tier values
const ModelTierSchema = z.enum(['LOCAL', 'NANO', 'MINI', 'FRONTIER', 'CUSTOM']);
export type ModelTier = z.infer<typeof ModelTierSchema>;

const ModelOptionsSchema = z.object({
  taskType: z.string(),
  stageId: z.number().optional(),
  modelTier: ModelTierSchema.optional(),
  governanceMode: z.enum(['DEMO', 'LIVE']).optional(),
  requirePhiCompliance: z.boolean().optional(),
  budgetLimit: z.number().positive().optional(),
  maxTokens: z.number().positive().optional(),
  temperature: z.number().min(0).max(2).optional(),
  streamResponse: z.boolean().optional(),
});

// Derive ModelOptions from the Zod schema — single source of truth
export type ModelOptions = z.infer<typeof ModelOptionsSchema>;

export interface LLMResponse {
  content: string;
  usage: {
    totalTokens: number;
    promptTokens: number;
    completionTokens: number;
  };
  cost: {
    total: number;
    input: number;
    output: number;
  };
  model: string;
  tier: string;
  finishReason: string;
  metadata?: Record<string, any>;
}

export interface BridgeConfig {
  orchestratorUrl: string;
  defaultTier: ModelTier;
  phiCompliantOnly: boolean;
  maxRetries: number;
  timeoutMs: number;
  costTrackingEnabled: boolean;
  streamingEnabled: boolean;
}

// Typed response from the AI Router routing endpoint
const RoutingApiResponseSchema = z.object({
  selectedTier: ModelTierSchema,
  model: z.string(),
  costEstimate: z.object({ total: z.number() }),
});

// ModelOptionsSchema and ModelOptions type are defined above the interfaces

const LLMResponseSchema = z.object({
  content: z.string(),
  usage: z.object({
    totalTokens: z.number(),
    promptTokens: z.number(),
    completionTokens: z.number(),
  }),
  cost: z.object({
    total: z.number(),
    input: z.number(),
    output: z.number(),
  }),
  model: z.string(),
  tier: z.string(),
  finishReason: z.string(),
  metadata: z.record(z.any()).optional(),
});

/**
 * AI Router Bridge for LangGraph Agents
 *
 * Handles communication between Python LangGraph agents and
 * the TypeScript AI Router service.
 */
export class AIRouterBridge {
  private config: BridgeConfig;
  private totalCost: number = 0;
  private requestCount: number = 0;

  constructor(config: Partial<BridgeConfig> = {}) {
    this.config = {
      orchestratorUrl: config.orchestratorUrl || 'http://localhost:3001',
      defaultTier: config.defaultTier || 'MINI',
      phiCompliantOnly: config.phiCompliantOnly ?? true,
      maxRetries: config.maxRetries || 3,
      timeoutMs: config.timeoutMs || 30000,
      costTrackingEnabled: config.costTrackingEnabled ?? true,
      streamingEnabled: config.streamingEnabled ?? false,
      ...config,
    };
  }

  /**
   * Single LLM invocation
   */
  async invoke(
    prompt: string,
    options: ModelOptions = { taskType: 'general' }
  ): Promise<LLMResponse> {
    const validatedOptions = ModelOptionsSchema.parse(options);

    // Step 1: Get routing recommendation
    const routingResponse = await this.getRouting(prompt, validatedOptions);

    // Step 2: Make LLM call with routing
    const llmResponse = await this.callLLM(prompt, {
      ...validatedOptions,
      modelTier: routingResponse.selectedTier,
      model: routingResponse.model,
    });

    // Step 3: Track costs if enabled
    if (this.config.costTrackingEnabled) {
      this.totalCost += llmResponse.cost.total;
      this.requestCount += 1;
    }

    return llmResponse;
  }

  /**
   * Streaming LLM invocation (generator)
   */
  async* streamInvoke(
    prompt: string,
    options: ModelOptions = { taskType: 'general' }
  ): AsyncGenerator<string, LLMResponse, unknown> {
    if (!this.config.streamingEnabled) {
      throw new Error('Streaming is disabled in bridge configuration');
    }

    const validatedOptions = ModelOptionsSchema.parse(options);

    // Get routing first
    const routingResponse = await this.getRouting(prompt, validatedOptions);

    // Create streaming request
    const response = await axios.post(
      `${this.config.orchestratorUrl}/api/ai/stream`,
      {
        prompt,
        ...validatedOptions,
        modelTier: routingResponse.selectedTier,
        model: routingResponse.model,
        stream: true,
      },
      {
        timeout: this.config.timeoutMs,
        responseType: 'stream',
      }
    );

    let content = '';
    const chunks: string[] = [];

    // Process streaming response
    for await (const chunk of response.data) {
      const chunkStr = chunk.toString();
      
      if (chunkStr.startsWith('data: ')) {
        const data = chunkStr.slice(6);
        
        if (data === '[DONE]') {
          break;
        }

        try {
          const parsed = JSON.parse(data);
          if (parsed.content) {
            chunks.push(parsed.content);
            content += parsed.content;
            yield parsed.content;
          }
        } catch (e) {
          // Skip invalid JSON chunks
        }
      }
    }

    // Return final response
    return {
      content,
      usage: {
        totalTokens: Math.ceil(content.length / 4), // Rough estimate
        promptTokens: Math.ceil(prompt.length / 4),
        completionTokens: Math.ceil(content.length / 4),
      },
      cost: {
        total: 0, // Calculated later
        input: 0,
        output: 0,
      },
      model: routingResponse.model,
      tier: routingResponse.selectedTier,
      finishReason: 'stop',
    };
  }

  /**
   * Batch LLM invocation
   */
  async batchInvoke(
    prompts: string[],
    options: ModelOptions = { taskType: 'general' }
  ): Promise<LLMResponse[]> {
    const validatedOptions = ModelOptionsSchema.parse(options);

    // Process prompts in parallel (with concurrency limit)
    const batchSize = 5; // Limit concurrent requests
    const results: LLMResponse[] = [];

    for (let i = 0; i < prompts.length; i += batchSize) {
      const batch = prompts.slice(i, i + batchSize);
      const batchPromises = batch.map((prompt) =>
        this.invoke(prompt, options)
      );

      const batchResults = await Promise.allSettled(batchPromises);

      for (const result of batchResults) {
        if (result.status === 'fulfilled') {
          results.push(result.value);
        } else {
          // Handle failed requests
          results.push({
            content: 'Error: Request failed',
            usage: { totalTokens: 0, promptTokens: 0, completionTokens: 0 },
            cost: { total: 0, input: 0, output: 0 },
            model: 'unknown',
            tier: 'unknown',
            finishReason: 'error',
            metadata: { error: result.reason },
          });
        }
      }
    }

    return results;
  }

  /**
   * Get cost statistics
   */
  getCostStats(): {
    totalCost: number;
    requestCount: number;
    averageCostPerRequest: number;
  } {
    return {
      totalCost: this.totalCost,
      requestCount: this.requestCount,
      averageCostPerRequest:
        this.requestCount > 0 ? this.totalCost / this.requestCount : 0,
    };
  }

  /**
   * Reset cost tracking
   */
  resetCostTracking(): void {
    this.totalCost = 0;
    this.requestCount = 0;
  }

  /**
   * Health check
   */
  async healthCheck(): Promise<{
    status: 'healthy' | 'unhealthy';
    orchestratorReachable: boolean;
    aiRouterAvailable: boolean;
    latencyMs: number;
  }> {
    const startTime = Date.now();

    try {
      const response = await axios.get(
        `${this.config.orchestratorUrl}/health`,
        { timeout: 5000 }
      );

      const latency = Date.now() - startTime;

      return {
        status: response.status === 200 ? 'healthy' : 'unhealthy',
        orchestratorReachable: true,
        aiRouterAvailable: response.data?.services?.ai_router === 'healthy',
        latencyMs: latency,
      };
    } catch (error) {
      return {
        status: 'unhealthy',
        orchestratorReachable: false,
        aiRouterAvailable: false,
        latencyMs: Date.now() - startTime,
      };
    }
  }

  // =====================================================================
  // Private Methods
  // =====================================================================

  /**
   * Get AI Router routing recommendation
   */
  private async getRouting(
    prompt: string,
    options: ModelOptions
  ): Promise<{
    selectedTier: ModelTier;
    model: string;
    estimatedCost: number;
  }> {
    const estimatedInputTokens = Math.ceil(prompt.length / 4); // Rough estimate
    const estimatedOutputTokens = Math.ceil(estimatedInputTokens * 0.5);

    try {
      const response = await axios.post(
        `${this.config.orchestratorUrl}/api/ai/router/route`,
        {
          taskType: options.taskType,
          estimatedInputTokens,
          estimatedOutputTokens,
          governanceMode: options.governanceMode || 'DEMO',
          preferredTier: options.modelTier || this.config.defaultTier,
          budgetLimit: options.budgetLimit,
          requirePhiCompliance: options.requirePhiCompliance ?? this.config.phiCompliantOnly,
          stageId: options.stageId,
        },
        { timeout: this.config.timeoutMs }
      );

      const parsed = RoutingApiResponseSchema.parse(response.data);
      return {
        selectedTier: parsed.selectedTier,
        model: parsed.model,
        estimatedCost: parsed.costEstimate.total,
      };
    } catch (error) {
      console.warn('[AIRouterBridge] Routing failed, using defaults:', error);
      
      // Fallback to default routing
      return {
        selectedTier: this.config.defaultTier,
        model: 'claude-3-5-sonnet-20241022', // Default fallback
        estimatedCost: 0.01, // Conservative estimate
      };
    }
  }

  /**
   * Make actual LLM call
   */
  private async callLLM(
    prompt: string,
    options: ModelOptions & { model?: string }
  ): Promise<LLMResponse> {
    let lastError: Error | undefined;

    for (let attempt = 1; attempt <= this.config.maxRetries; attempt++) {
      try {
        const response = await axios.post(
          `${this.config.orchestratorUrl}/api/ai/invoke`,
          {
            prompt,
            model: options.model,
            modelTier: options.modelTier,
            taskType: options.taskType,
            maxTokens: options.maxTokens || 4096,
            temperature: options.temperature || 0.7,
            governanceMode: options.governanceMode || 'DEMO',
            stageId: options.stageId,
          },
          { timeout: this.config.timeoutMs }
        );

        // Validate and return response
        const validated: LLMResponse = LLMResponseSchema.parse(response.data) as LLMResponse;
        return validated;
      } catch (error) {
        lastError = error as Error;
        
        if (error instanceof axios.AxiosError) {
          // Don't retry on client errors (4xx)
          if (error.response?.status && error.response.status < 500) {
            throw error;
          }
        }

        // Wait before retry (exponential backoff)
        if (attempt < this.config.maxRetries) {
          await new Promise((resolve) =>
            setTimeout(resolve, Math.pow(2, attempt) * 1000)
          );
        }
      }
    }

    throw lastError || new Error('Max retries exceeded');
  }
}

/**
 * Factory function to create AI Router Bridge from environment
 */
export function createAIRouterBridge(
  config: Partial<BridgeConfig> = {}
): AIRouterBridge {
  const parsedTier = ModelTierSchema.safeParse(process.env.AI_DEFAULT_TIER);
  const envConfig: Partial<BridgeConfig> = {
    orchestratorUrl: process.env.AI_ROUTER_URL || 'http://localhost:3001',
    defaultTier: parsedTier.success ? parsedTier.data : 'MINI',
    phiCompliantOnly: process.env.PHI_COMPLIANT_ONLY === 'true',
    maxRetries: parseInt(process.env.AI_MAX_RETRIES || '3'),
    timeoutMs: parseInt(process.env.AI_TIMEOUT_MS || '30000'),
    costTrackingEnabled: process.env.AI_COST_TRACKING !== 'false',
    streamingEnabled: process.env.AI_STREAMING_ENABLED === 'true',
  };

  return new AIRouterBridge({ ...envConfig, ...config });
}

// Default export
export default AIRouterBridge;